lib/msf/core/exploit/browser_autopwnv2.rb

```diff
@@ -0,0 +1,834 @@
+    # @return [void]
+    def log_click(ip, data='')
+      report_note(
+        :host => ip,
+        :type => 'bap.clicks',
+        :data => data,
+        :update => :unique)
+    end
+
+
+    # Prints a list of suitable exploits for the current list.
+    #
+    # @see #sort_bap_exploits Explains how the exploit list is generated at first.
+    # @see #get_suitable_exploits Explains how we serve exploits to each client.
+    # @return [void]
+    def show_real_list(ip, tag, current_exploit_list)
+      order = 1
+      table = Rex::Ui::Text::Table.new(
+        'Header'  => '',
+        'Indent'  => 1,
+        'Columns' => ['Order', 'IP', 'Exploit']
+      )
+      current_exploit_list.each do |m|
+        table << [order, ip, m.shortname]
+        order += 1
+      end
+
+      if table.rows.empty?
+        print_status("User #{cli.peerhost} (Tag: #{tag}) visited our malicious link, but no exploits found suitable.")
+      else
+        # Update the exploit list data
+        log_click(cli.peerhost, table.to_csv)
+        print_status("Exploits found suitable for #{cli.peerhost} (Tag: #{tag})#{table}")
+      end
+    end
+
+
+    # Returns a list of exploit URLs. This is used by #build_html so the client can load our
+    # exploits one by one.
+    #
+    # @see #build_html
+    # @param cli [Socket] Socket for the browser
+    # @param request [Rex::Proto::Http::Request] The HTTP request sent by the browser
+    # @return [Array]
+    def get_exploit_urls(cli, request)
+      urls = []
+
+      exploit_list = get_suitable_exploits(cli, request)
+
+      exploit_list.each do |mod|
+        proto = datastore['SSL'] ? 'https' : 'http'
+        host = datastore['URIHOST'] || Rex::Socket.source_address
+        port = datastore['SRVPORT']
+        resource = mod.datastore['URIPATH']
+        url = "#{proto}://#{host}:#{port}#{resource}"
+        urls << url
+      end
+
+      urls
+    end
+
+
+    # Handles client requests specific for BAP.
+    #
+    # @param cli [Socket] Socket for the browser
+    # @param request [Rex::Proto::Http::Request] The HTTP request sent by the browser
+    # @return [void]
+    def on_request_uri(cli, request)
+      # Check if target is on our whitelist
+      if @whitelist && !is_ip_targeted?(cli.peerhost)
+        print_status("Client is trying to connect but not on our whitelist.")
+        send_not_found(cli)
+        return
+      end
+
+      log_click(cli.peerhost)
+
+      super
+    end
+
+
+    # Returns true if the IP is on our whitelist.
+    #
+    # @param [String] cli_ip Client's IP.
+    # @return [TrueClass] The IP is on the whitelist.
+    # @return [FalseClass] The IP is not on the whitelist.
+    def is_ip_targeted?(cli_ip)
+      return true unless @whitelist
+      @whitelist.include?(cli_ip)
+    end
+
+
+    # Returns a number of sessions obtained by BAP's payload handlers.
+    #
+    # @return [Fixnum] A session count.
+    def session_count
```

Comment:
Hey @hmoore-r7 could you please explain this a little bit?

I tried to reproduce this scenario but I can't:

```
msf > workspace -a 1
[*] Added workspace: 1
msf > use exploit/multi/handler 
msf exploit(handler) > set lport 5555
lport => 5555
msf exploit(handler) > set exitonsession false
exitonsession => false
msf exploit(handler) > run -j
[*] Exploit running as background job.

[*] Started reverse handler on 192.168.1.64:5555 
msf exploit(handler) > [*] Starting the payload handler...

msf exploit(handler) > workspace -a 2
[*] Added workspace: 2
msf exploit(handler) > set lport 4444
lport => 4444
msf exploit(handler) > run -j
[*] Exploit running as background job.

[*] Started reverse handler on 192.168.1.64:4444 
msf exploit(handler) > [*] Starting the payload handler...

msf exploit(handler) > jobs


Jobs
====

  Id  Name
  --  ----
  0   Exploit: multi/handler
  1   Exploit: multi/handler
```

It looks like jobs aren't workspace-specific, they're always global?